File: onyx-remote-database/src/main/java/com/onyx/application/DatabaseServer.java
package com.onyx.application;

import com.onyx.client.SSLPeer;
import com.onyx.client.auth.AuthenticationManager;
import com.onyx.entity.SystemUser;
import com.onyx.entity.SystemUserRole;
import com.onyx.persistence.context.impl.ServerSchemaContext;
import com.onyx.persistence.factory.PersistenceManagerFactory;
import com.onyx.factory.impl.ServerPersistenceManagerFactory;
import com.onyx.persistence.manager.PersistenceManager;
import com.onyx.server.auth.DefaultAuthenticationManager;
import com.onyx.server.base.AbstractDatabaseServer;
import com.onyx.server.base.ServerState;
import com.onyx.server.cli.CommandLineParser;
import com.onyx.server.rmi.OnyxRMIServer;
import com.onyx.interactors.encryption.DefaultEncryptionInteractor;
import com.onyx.interactors.encryption.EncryptionInteractor;
import org.jetbrains.annotations.NotNull;

/**
 * Base Database Server Application.
 * <p>
 * All servers must inherit from this class in order to extend the Onyx Remote Database.  This does not have
 * any logic that is used within the Web Database.
 * <p>
 * <pre>
 * <code>
 *
 *    DatabaseServer server1 = new DatabaseServer();
 *    server1.setPort(8080);
 *    server1.setDatabaseLocation("C:/Sandbox/Onyx/Tests/server.oxd");
 *    server1.start();
 *    server1.join();
 *
 *
 *   To invoke command line
 *   java -cp path_to_my_classes(.jar):onyx-remote-database-(0.0.1)-alpha.jar com.onyx.application.DatabaseServer -l /path/to/database/location -port=8080
 *
 *   e.x
 *   java -cp /Users/tosborn1/Dropbox/OnyxSandbox/onyxdb-parent/OnyxDatabaseTests/target/classes/:onyx-remote-database-0.0.1-alpha.jar com.onyx.application.DatabaseServer -l /Users/tosborn1/Desktop/database1 -port=8080
 * </code>
 * </pre>
 *
 * @author Tim Osborn
 * @since 1.0.0
 */
public class DatabaseServer extends AbstractDatabaseServer implements OnyxServer {

    @SuppressWarnings("WeakerAccess")
    public static final String PERSISTENCE_MANAGER_SERVICE = "1";
    @SuppressWarnings("WeakerAccess")
    public static final String AUTHENTICATION_MANAGER_SERVICE = "2";

    // RMI Server.  This is the underlying network io server
    @SuppressWarnings("WeakerAccess")
    protected OnyxRMIServer rmiServer;

    @SuppressWarnings("WeakerAccess")
    protected PersistenceManagerFactory persistenceManagerFactory;

    @SuppressWarnings("WeakerAccess")
    protected AuthenticationManager authenticationManager = null;

    private EncryptionInteractor encryption = DefaultEncryptionInteractor.INSTANCE;

    /**
     * Constructor
     *
     * @since 1.3.0 Broke out a new ServerPersistenceManager to handle
     *              the differences in query caching
     */
    public DatabaseServer() {
    }

    @SuppressWarnings("unused")
    public DatabaseServer(boolean avoidDefaultConstructor) {

    }

    /**
     * Run Database Server Main Method
     * <p>
     * ex:  executable /Database/Location/On/Disk 8080 admin admin
     *
     * @param args Command Line Arguments
     * @throws Exception General Exception
     * @since 1.0.0
     */
    public static void main(String[] args) throws Exception {

        final DatabaseServer instance = new DatabaseServer();

        final CommandLineParser commandLineParser = new CommandLineParser();
        commandLineParser.configureDatabaseWithCommandLineOptions(instance, args);

        instance.start();
        instance.join();
    }


    /**
     * Start the database socket server
     *
     * @since 1.0.0
     */
    @Override
    public void start() {
        if (state != ServerState.RUNNING) {
            try {
                this.persistenceManagerFactory = new ServerPersistenceManagerFactory(this.location);
                this.persistenceManagerFactory.setCredentials(this.user, this.password);
                this.persistenceManagerFactory.initialize();

                // Create a default user
                SystemUser user = new SystemUser();
                user.setUsername(this.user);
                user.setPassword(encryption.encrypt(this.password));
                user.setRole(SystemUserRole.ROLE_ADMIN);

                // Default User and password
                persistenceManagerFactory.getPersistenceManager().saveEntity(user);

                this.authenticationManager = new DefaultAuthenticationManager(persistenceManagerFactory.getPersistenceManager(), encryption);

                // Create the RMI Server
                rmiServer = new OnyxRMIServer();
                rmiServer.setPort(this.port);
                rmiServer.setSslKeystoreFilePath(this.sslKeystoreFilePath);
                rmiServer.setSslKeystorePassword(this.sslKeystorePassword);
                rmiServer.setSslTrustStoreFilePath(this.sslTrustStoreFilePath);
                rmiServer.setSslTrustStorePassword(this.sslTrustStorePassword);
                rmiServer.setSslStorePassword(this.sslStorePassword);
                rmiServer.setMaxWorkerThreads(this.maxWorkerThreads);

                registerServices();

                try {
                    rmiServer.start();
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }

                // Set the push publisher within the schema context so the query caching mechanism can
                // tell what push clients to send updates to.
                ((ServerSchemaContext)this.persistenceManagerFactory.getSchemaContext()).setPushPublisher(rmiServer);

                state = ServerState.RUNNING;

            } catch (Throwable t) {
                t.printStackTrace(System.err);
            }
        }
    }

    /**
     * Register services.  This method registers all of the proxy objects and makes them public
     */
    @SuppressWarnings("WeakerAccess")
    protected void registerServices()
    {
        // Register the Persistence Manager
        rmiServer.register(PERSISTENCE_MANAGER_SERVICE, this.persistenceManagerFactory.getPersistenceManager(), PersistenceManager.class);
        rmiServer.register(AUTHENTICATION_MANAGER_SERVICE, this.authenticationManager, AuthenticationManager.class);
    }

    /**
     * Stop the database server
     *
     * @since 1.0.0
     */
    public void stop()
    {
        rmiServer.stop();
        if (persistenceManagerFactory != null) {
            persistenceManagerFactory.close();
        }
        super.stop();
    }

    /**
     * Get persistence manager
     *
     * @return the underlying persistence manager
     * @since 1.2.3
     */
    @SuppressWarnings("unused")
    public PersistenceManager getPersistenceManager() {
        return this.persistenceManagerFactory.getPersistenceManager();
    }

    public void copySSLPeerTo(SSLPeer peer) {

    }

    @NotNull
    @Override
    public EncryptionInteractor getEncryption() {
        return encryption;
    }

    @Override
    public void setEncryption(@NotNull EncryptionInteractor encryptionInteractor) {
        this.encryption = encryptionInteractor;
    }
}

File: onyx-remote-database/src/main/kotlin/com/onyx/application/impl/DatabaseServer.kt
package com.onyx.application.impl

import com.onyx.application.OnyxServer
import com.onyx.network.auth.AuthenticationManager
import com.onyx.entity.SystemUser
import com.onyx.entity.SystemUserRole
import com.onyx.persistence.context.impl.ServerSchemaContext
import com.onyx.persistence.factory.PersistenceManagerFactory
import com.onyx.persistence.factory.impl.ServerPersistenceManagerFactory
import com.onyx.persistence.manager.PersistenceManager
import com.onyx.network.auth.impl.DefaultAuthenticationManager
import com.onyx.cli.CommandLineParser
import com.onyx.network.rmi.OnyxRMIServer
import com.onyx.interactors.encryption.impl.DefaultEncryptionInteractorInstance
import com.onyx.interactors.encryption.EncryptionInteractor
import com.onyx.persistence.IManagedEntity
import com.onyx.diskmap.store.StoreType

/**
 * Base Database Server Application.
 *
 *
 * All servers must inherit from this class in order to extend the Onyx Remote Database.  This does not have
 * any logic that is used within the Web Database.
 *
 *
 *
 * DatabaseServer server1 = new DatabaseServer();
 * server1.setPort(8080);
 * server1.setDatabaseLocation("C:/Sandbox/Onyx/Tests/server.oxd");
 * server1.start();
 * server1.join();
 *
 *
 * To invoke command line
 * java -cp path_to_my_classes(.jar):onyx-remote-database-(0.0.1)-alpha.jar com.onyx.application.impl.DatabaseServer -l /path/to/database/databaseLocation -port=8080
 *
 * e.x
 * java -cp /Users/Tim Osborn/Dropbox/OnyxSandbox/onyxdb-parent/OnyxDatabaseTests/target/classes/:onyx-remote-database-0.0.1-alpha.jar com.onyx.application.impl.DatabaseServer -l /Users/Tim Osborn/Desktop/database1 -port=8080
 *
 *
 * @author Tim Osborn
 * @since 1.0.0
 */
open class DatabaseServer(override val databaseLocation:String) : AbstractDatabaseServer(databaseLocation), OnyxServer {

    override var encryption: EncryptionInteractor = DefaultEncryptionInteractorInstance

    // RMI Server.  This is the underlying network io server
    @Suppress("MemberVisibilityCanPrivate")
    protected lateinit var rmiServer: OnyxRMIServer

    open var persistenceManagerFactory: PersistenceManagerFactory? = null
    @Suppress("MemberVisibilityCanPrivate")
    protected var authenticationManager: AuthenticationManager? = null

    var storeType: StoreType = StoreType.FILE

    /**
     * Start the database socket server
     *
     * @since 1.0.0
     */
    override fun start() {
        if (!isRunning) {
            if (this.persistenceManagerFactory == null) {
                this.persistenceManagerFactory = ServerPersistenceManagerFactory(this.databaseLocation)
                this.persistenceManagerFactory?.setCredentials(this.user, this.password)
                this.persistenceManagerFactory?.storeType = this.storeType
                this.persistenceManagerFactory?.initialize()
            }

            // Create a default user
            val user = SystemUser()
            user.username = this.user
            user.password = encryption.encrypt(this.password)
            user.role = SystemUserRole.ROLE_ADMIN
            user.firstName = "Admin"
            user.lastName = "Admin"

            // Default User and password
            this.persistenceManagerFactory?.persistenceManager?.saveEntity<IManagedEntity>(user)
            this.authenticationManager = DefaultAuthenticationManager(persistenceManagerFactory!!.persistenceManager, encryption)

            // Create the RMI Server
            this.rmiServer = OnyxRMIServer()
            this.rmiServer.port = port
            this.registerServices()

            this.rmiServer.start()

            // Set the push publisher within the schema context so the query caching mechanism can
            // tell what push clients to send updates to.
            (this.persistenceManagerFactory?.schemaContext as ServerSchemaContext).setPushPublisher(rmiServer)

            isRunning = true
        }
    }

    /**
     * Register services.  This method registers all of the proxy objects and makes them public
     */
    @Suppress("MemberVisibilityCanPrivate")
    protected fun registerServices() {
        // Register the Persistence Manager
        rmiServer.register(PERSISTENCE_MANAGER_SERVICE, this.persistenceManagerFactory!!.persistenceManager, PersistenceManager::class.java)
        rmiServer.register(AUTHENTICATION_MANAGER_SERVICE, this.authenticationManager!!, AuthenticationManager::class.java)
    }

    /**
     * Stop the database server
     *
     * @since 1.0.0
     */
    override fun stop() {
        rmiServer.stop()
        if (persistenceManagerFactory != null) {
            persistenceManagerFactory!!.close()
        }
    }

    /**
     * Join Server.  Have it suspend on a daemon thread
     *
     * @since 1.2.0
     */
    override fun join() = rmiServer.join()

    /**
     * Get persistence manager
     *
     * @return the underlying persistence manager
     * @since 1.2.3
     */
    val persistenceManager: PersistenceManager
        get() = this.persistenceManagerFactory!!.persistenceManager

    /**
     * Register services.  This method registers all of the proxy objects and makes them public
     */
    @Suppress("MemberVisibilityCanPrivate")
    fun <T> register(
        name: String,
        implementation: T,
        type: Class<T>,
    ) {
        rmiServer.register(name, implementation as Any, type)
    }

    /**
     * Register services.  This method registers all of the proxy objects and makes them public
     */
    @Suppress("MemberVisibilityCanPrivate")
    fun <T> unregister(
        name: String,
    ) {
        rmiServer.deregister(name)
    }

    companion object {

        const val PERSISTENCE_MANAGER_SERVICE = "1"
        const val AUTHENTICATION_MANAGER_SERVICE = "2"

        /**
         * Run Database Server Main Method
         *
         *
         * ex:  executable /Database/Location/On/Disk 8080 admin admin
         *
         * @param args Command Line Arguments
         * @throws Exception General Exception
         * @since 1.0.0
         */
        @Suppress("NON_FINAL_MEMBER_IN_OBJECT")
        @Throws(Exception::class)
        @JvmStatic
        open fun main(args: Array<String>) {

            val commandLineParser = CommandLineParser(args)
            val instance = DatabaseServer(commandLineParser.databaseLocation)
            commandLineParser.configureDatabaseWithCommandLineOptions(instance)

            instance.start()
            instance.join()
        }

    }
}


